File: src/AI/NPC_AI_System/States/DefensiveState.cs
public class DefensiveState : INPCState
{
    public void Enter(NPCController npc)
    {
        GD.Print($"{npc.Name} is now in Defensive mode due to extreme weather!");
        // Increase defense or take other defensive measures
        npc.Defense += 20f;
    }

    public void Execute(NPCController npc, float deltaTime)
    {
        // Defensive behavior logic during extreme weather
        // Example: Move to designated shelter city
        if (npc.CurrentCity.Name != "Shelter")
        {
            City shelterCity = npc.GameManager.FindShelter();
            npc.MoveToCity(shelterCity);
            GD.Print($"{npc.Name} is seeking shelter...");
        }
    }

    public void Exit(NPCController npc)
    {
        GD.Print($"{npc.Name} exits Defensive mode.");
        npc.Defense -= 20f;
    }
}

File: src/AI/States/DefensiveState.cpp
#include "DefensiveState.hpp"
#include "../NPCController.hpp"
#include <godot_cpp/core/class_db.hpp>

void DefensiveState::enter(NPCController* npc) {
    godot::UtilityFunctions::print(godot::String("{0} is now in Defensive mode!").format(Array::make(npc->get_name())));
    npc->set_defense(npc->get_defense() + 20.0f);
}

void DefensiveState::execute(NPCController* npc, float deltaTime) {
    // Implement defensive behavior
    // Example: Check for threats, move to safe locations, etc.
}

void DefensiveState::exit(NPCController* npc) {
    godot::UtilityFunctions::print(godot::String("{0} exits Defensive mode.").format(Array::make(npc->get_name())));
    npc->set_defense(npc->get_defense() - 20.0f);
} 